### PR TITLE
refactor: move `aggregate_statistics` to `datafusion-physical-optimizer`

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Parts #11502 

## Rationale for this change

- move `aggregate_statistics` to `datafusion-physical-optimizer`
- leave the tests to `datafusion/core/tests/physical_optimizer_integration.rs`

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/tests/aggregate_statistics.rs [hunk 1]
```diff
@@ -15,292 +15,19 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! Utilizing exact statistics from sources to avoid scanning data
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/tests/aggregate_statistics.rs:18]: I worry that putting these tests  in `datafusion/core/src/physical_optimizer/tests/aggregate_statistics.rs` will make them harder to find becase:
1. It is in a different crate than being tested
2. It is in a non standard testing location

As I understand it, these tests use code in other modules that physical optimizer doesn't depend on (like `functions-aggregate`


Thus, I would like to suggest we put these tests https://github.com/apache/datafusion/tree/main/datafusion/core/tests


Perhaps something like `physical_optimizer_integration.rs` to mirror https://github.com/apache/datafusion/blob/main/datafusion/core/tests/optimizer_integration.rs

**Review by alamb — COMMENTED** [review]
Thank you very much @Weijun-H  -- I think you are doing the hard work of figuring out the patterns as we start to break out the physical optimizers from the core

In this case, I think the challenge is where to put the tests for `physical-optimizer` passes that rely on code that is not available to the `physical-optimizer` crate (like the functions library or the parquet exec)

We had a similar challenge with the logical optimizer passes 

I think the two main options are:
1. Move the tests into "integration" tests in https://github.com/apache/datafusion/tree/main/datafusion/core/tests
2. Rewrite the tests to avoid the dependencies (e.g. use [function stubs](https://github.com/apache/datafusion/blob/main/datafusion/expr/src/test/function_stub.rs) and other things

I personally suggest we move the tests to https://github.com/apache/datafusion/tree/main/datafusion/core/tests so they retain the same level of coverage

The downside of this is that then the tests will not be in the same directory as code. I think we can manage this with some good comments / documentation (e.g. at the end of the physical optimizer module)


cc @lewiszlw @mustafasrepo @jayzhan211

File: datafusion/common/src/utils/expr.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+//! Expression utilities
+
+use crate::ScalarValue;
+
+///  The value to which `COUNT(*)` is expanded to in
+///  `COUNT(<constant>)` expressions
+pub const COUNT_STAR_EXPANSION: ScalarValue = ScalarValue::Int64(Some(1));
```
Comment (Weijun-H) [inline datafusion/common/src/utils/expr.rs:24]: move `COUNT_STAR_EXPANSION` here

File: datafusion/core/src/test_util/mod.rs [hunk 3]
```diff
@@ -402,3 +407,57 @@ pub fn bounded_stream(batch: RecordBatch, limit: usize) -> SendableRecordBatchSt
...
 }
+
+/// Describe the type of aggregate being tested
+pub enum TestAggregate {
+    /// Testing COUNT(*) type aggregates
+    CountStar,
+
+    /// Testing for COUNT(column) aggregate
+    ColumnA(Arc<Schema>),
+}
```
Comment (Weijun-H) [inline datafusion/core/src/test_util/mod.rs:418]: This enum is used by `aggregate_statistics` and `limited_distinct_aggregate`, so move it for convenience.

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H  -- this looks great.

**Pull request merged**